src/libopensc/card-mcrd.c

```diff
@@ -1391,6 +1395,72 @@ static int mcrd_compute_signature(sc_card_t * card,
 	SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, apdu.resplen);
 }
 
+static int mcrd_decipher(struct sc_card *card,
+						 const u8 * crgram, size_t crgram_len,
+						 u8 * out, size_t outlen)
+{
+	sc_security_env_t *env = NULL;
+	int r = 0;
+	size_t sbuf_len = 0, tags1_len = 0, tags2_len = 0;
+	sc_apdu_t apdu;
+	u8 *sbuf = NULL, *p = NULL;
+
+	if (card == NULL || crgram == NULL || out == NULL)
+		return SC_ERROR_INVALID_ARGUMENTS;
+	env = &DRVDATA(card)->sec_env;
+
+	LOG_FUNC_CALLED(card->ctx);
+	if (env->operation != SC_SEC_OPERATION_DERIVE)
+		SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, iso_ops->decipher(card, crgram, crgram_len, out, outlen));
+	if (crgram_len > 255)
+		SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, SC_ERROR_INVALID_ARGUMENTS);
+
+	sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL,
+		 "Will dervie (%d) for %"SC_FORMAT_LEN_SIZE_T"u (0x%02"SC_FORMAT_LEN_SIZE_T"x) bytes using key %d algorithm %d flags %d\n",
+		 env->operation, crgram_len, crgram_len, env->key_ref[0],
+		 env->algorithm, env->algorithm_flags);
+
+	// Calculate length of buffer
+	tags1_len = sc_asn1_put_tag(0x86, NULL, crgram_len, NULL, 0, NULL);
+	LOG_TEST_RET(card->ctx, tags1_len, "Error handling TLV.");
+	tags2_len = sc_asn1_put_tag(0x7F49, NULL, tags1_len, NULL, 0, NULL);
+	LOG_TEST_RET(card->ctx, tags2_len, "Error handling TLV.");
+	sbuf_len = sc_asn1_put_tag(0xA6, NULL, tags2_len, NULL, 0, NULL);
+	LOG_TEST_RET(card->ctx, sbuf_len, "Error handling TLV.");
```

Comment:
I thought more of something like https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/pkcs15-sc-hsm.c#L149, i.e. define specific ASN.1 structures, to make the structure more readable...